editoast/src/core/mod.rs

```
@@ -456,10 +435,13 @@ mod tests {
             .response(StatusCode::NOT_FOUND)
             .body(error.to_string())
             .finish();
-        let mut error_with_status: InternalError = serde_json::from_value(error).unwrap();
-        error_with_status.set_status(StatusCode::NOT_FOUND);
+        let expected_error: StandardCoreError = serde_json::from_value(error).unwrap();
         let result = Req.fetch(&core.into()).await;
-        let expected_err: InternalError = error_with_status;
-        assert_eq!(result, Err(expected_err));
+        let result = result.err().unwrap();
+        if let Error::StandardCoreError(standard_core_error) = result {
+            assert_eq!(standard_core_error, expected_error);
+        } else {
+            panic!("Unexpected error type: {:?}", result);
+        }
```

Nit: using assert_eq with the `Error` variant would be more idiomatic than if-else panicking imo

---

hint: `Error` does not implement `PartialEq` which I suppose led to this version of the code. Sure, you could add `#[cfg_attr(test, derive(PartialEq))]`, but you could also do the following I believe (not tested).

```rust
assert!(matches!(result, Error::StandardCoreError(expected_error)));
```

---

I'm fine with both, with a slight preference for `assert_eq` still as it will show the actual value on failure unlike `matches!` which will consume it.

---

I can't implement `PartialEq` because `lapin::Error` and `serde_json::Error` inside `Error::MqClientError` don't implement this trait.

---

Oh ok, then I'd say that `matches!` is our best option then

---

The `match` only checks the type of the variant (`Error::StandardCoreError`) and ignores the error's content `expected_error`. Is this what we want here?

---

Oh, right, with this syntax, it must bind to a new name `expected_error`. So maybe you can do it like this?

```rust
assert!(matches!(result, Error::StandardCoreError(standard_core_error) if standard_core_error == expected_error));
```

---

Yeah I agree the match expression isn't great, it kinda looks like perl... About PartialEq you could use derivative to ignore lapin::Error: https://mcarton.github.io/rust-derivative/latest/cmp.html

---

Thank you for the suggestion, but strangely, it only works for struct and not for enum.
```rust
#[derive(Derivative)]
#[derivative(PartialEq)]
struct Foo {
    foo: u8,
    #[derivative(PartialEq = "ignore")]
    bar: u8,
}

#[derive(Derivative)]
#[derivative(PartialEq)]
enum Foo {
    Foo,
    #[derivative(PartialEq = "ignore")] // error: Unknown attribute `ignore` for trait `PartialEq`
    Bar,
}
```

---

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=9755b45b30128f52be01c75848a5208c

---

I looks like `derivative` has become unmaintained :/ 

https://github.com/OpenRailAssociation/osrd/issues/11071

---

Thanks for noticing that. In the meantime, should I keep the current version?

---

I'd suggest to still use derivative in this precise case, as it looks like we can just `sed/derivative/educe` so that creates little additional work.

---

Done.